Allow WooPay to request session data from merchant

bborman22 opened the pull request:
#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->
This PR is to address a performance optimization in the WooPay Direct Checkout feature. This PR is one half of the fix for https://github.com/Automattic/woopay/issues/2480. This PR should be tested in tandem with https://github.com/Automattic/woopay/pull/2529.

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

The overall feature is to send a minimum amount of session data to WooPay to start, then let WooPay request the rest of the session data only if it is needed (i.e. the shopper is already logged into WooPay). If the shopper is not logged into WooPay the full session data should not be requested and the shopper should be taken to the merchant checkout page. More details about the feature and this optimization can be found in https://github.com/Automattic/woopay/issues/2480.

To accomplish that goal, this PR adds the following:
- The ability to retrieve the minimum session data either on page load or through AJAX if necessary.
- ~~Pass the minimum session data to the connect page and receive a redirect URL.~~
- Pass the encrypted and signed minimum session data to WooPay with a redirect.
- A new REST endpoint that allows WooPay to request the full session data from the merchant.

The points of failure I focused on in this PR:
- Inability to get the minimum session data.
- Inability to retrieve full session data from the merchant.

In all cases when there is a failure there, we should default to taking the shopper to the merchant checkout page.

**Notes**
One key thing to note is the need to unset the `preloaded_requests` on the full session data response. This is unfortunate as it reduces some of the performance benefits. The reason this had to be unset is that the preloaded requests use `rest_preload_api_request` which just replicates a REST request but with no headers, which means no `Cart-Token`. This leads the cart not being found and instead returning an empty cart object breaking things on WooPay. I experimented with other ways to request the cart data doing something like this:

```
add_filter( 'woocommerce_store_api_disable_nonce_check', '__return_true' );
$cart_request = new WP_REST_Request('GET', '/wc/store/v1/cart' );
$cart_request->set_headers( $woopay_request->get_headers() );
$cart_data = rest_do_request( $cart_request );

$checkout_request = new WP_REST_Request('GET', '/wc/store/v1/checkout' );
$checkout_request->set_headers( $woopay_request->get_headers() );
$checkout_data = rest_do_request( $checkout_request );
remove_filter( 'woocommerce_store_api_disable_nonce_check', '__return_true' );
```

But ran into the problem that the initial requests had been fired and so WooCommerce wasn't trying to load the cart again since it already loaded it on the first requests in `get_init_session_request` and it generally seemed "too late" to update the current session to use the Cart Token. **Given the time constraints I opted to leave the unset in for now and revisit in a separate issue, but if we find a good solution I'd love to add it to this PR!**

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

All tests should start with WooPay enabled, the `_wcpay_feature_woopay_direct_checkout` option set to 1 (enabled), and an item in the cart.

- When the shopper is logged out of WooPay, clicking the `Proceed to Checkout` button should take the shopper to the merchant checkout. If third party cookies are disabled, there will be a request to `woopay/?checkout_redirect=1&blog_id=...` that results in a 302 redirect to the merchant checkout. In this case nothing should ever get stored on WooPay for the session.
- When the shopper is logged in to WooPay,  clicking the `Proceed to Checkout` button should take the shopper to the WooPay checkout. If third party cookies are disabled, there will be a request to `woopay/?checkout_redirect=1&blog_id=...` that results in a 302 redirect to WooPay with a `platform_checkout_key`. 
- Test the existing flows are unaffected:
    - Using the WooPay Express checkout button on the cart and checkout page.
    - Using the OTP modal on checkout page after entering an email address.
    - Checking out without WooPay.
- Test in both blocks and shortcode cart.
- Test with a shopper that is logged into the merchant store.
- Test with the WooPay PR running on the staging sandbox to confirm environment compatibility. 

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

bborman22 committed:
Allow WooPay to request session data from merchant

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8268` or branch name `add/woopay-merchant-request` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/woopay-merchant-request) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: a265f797ff52cb9d96c8793262f3690b97e4f7f7
- Build time: 2024-03-21 22:44:09 UTC

_Note: the build is updated when a new commit is pushed to this PR._

bborman22 committed:
Send temporary session data through connect page

bborman22 committed:
Add error handling to cache session requests

bborman22 committed:
Merge changes with timeout changes

bborman22 committed:
Add URL validation to WooPay checkout URL

bborman22 committed:
Add permission check to session endpoint

bborman22 committed:
Rename minimum session data endpoints, optimize minimum session data retrieval

bborman22 committed:
Fix merge conflicts

bborman22 committed:
Merge branch 'develop' into add/woopay-merchant-request

bborman22 committed:
Add changelog

bborman22 committed:
Merge remote-tracking branch 'origin/add/woopay-merchant-request' into add/woopay-merchant-request

bborman22 committed:
Fix linting errors

bborman22 committed:
Fix psalm errors

bborman22 commented:
Adding @ricardo as a reviewer given the round robin assignment from the WooPay PR.

ricardo commented on client/checkout/woopay/direct-checkout/woopay-direct-checkout.js:
```diff
@@ -234,6 +279,51 @@ class WooPayDirectCheckout {
+	 * @return {Promise<Promise<*>|*>} Resolves to the WooPay redirect response.
+	 */
+	static async getWooPayMinimumSesssionDataFromMerchant() {
+		if ( getConfig( 'woopayMinimumSessionData' ) ) {
+			return getConfig( 'woopayMinimumSessionData' );
+		}
```
Shouldn't this always be defined from the cart page? Or are we making the request below just in case the user clicks the checkout link from a different page when this feature becomes available?

Just to understand if I'm missing something.

It would be nice to add some comment explaining when we expect this to not be available.

bborman22 replied:
Yeah, it should always be defined. I was just being extra cautious here with a request fallback in case of some edge cases. I've added a comment in https://github.com/Automattic/woocommerce-payments/pull/8268/commits/416f4d9c49cfe93f2999ab0f71becf4e3298862a to clarify that

ricardo commented on tests/e2e/specs/wcpay/merchant/merchant-orders-full-refund.spec.js:
```diff
@@ -33,6 +33,15 @@ describe( 'Order > Full refund', () => {
+		orderAmount = await page.evaluate(
+			( el ) => el.textContent,
+			priceElement
+		);
```
I assume this was an unrelated fix that we're including just because it's small? 🙂

bborman22 replied:
Doh! Totally unreletad change I was using to get some local E2E tests to pass. I've removed this change in https://github.com/Automattic/woocommerce-payments/pull/8268/commits/dfc8581500df294c4d98f456682c9b341f5e9866 as it has nothing to do with the PR, sorry about that!

ricardo commented on includes/admin/class-wc-rest-woopay-session-controller.php:
```diff
@@ -0,0 +1,104 @@
+ * It is only available here as a temporary workaround for the issue described in
+ * https://github.com/Automattic/woocommerce-payments/issues/6304
+ *
+ * This file can be removed once the issue is resolved.
```
- This endpoint is in fact registered, right?
- Line 3 should read `WC_REST_WooPay_Session_Controller`.

bborman22 replied:
This was a copy/paste issue and updated it in https://github.com/Automattic/woocommerce-payments/pull/8268/commits/da4c85f5f2deabca9b4fb2c2fed2a7478529ca64.

ricardo commented on includes/admin/class-wc-rest-woopay-session-controller.php:
```diff
@@ -0,0 +1,104 @@
+	 * Configure REST API routes.
+	 */
+	public function register_routes() {
+		// TODO: We need to give this proper permission checks, required fields, etc.
```
Do you have anything else in mind for the permission check other than `has_valid_request_signature`?

It might be a good idea to add the required fields 👍.

bborman22 replied:
That todo was outdated at this point. There aren't really any required fields on this request (at this point), though it relies on some headers. Those headers are then checked in the `is_request_from_woopay` and `has_valid_request_signature` functions so the fix here was to just remove the todo, but let me know if you see another opportunity to check these permissions.

ricardo reviewed (changes requested):
Dropped the main comment in the WooPay repo, but the approach looks good to me. Nice work!

Only left pretty minor questions on this PR.

All tests have passed:

✅ With third party cookies disabled, there's a request to woopay/?cache_checkout_key that redirects to the merchant checkout if the user is logged out.
✅ With third party cookies disabled, there's a request to woopay/?cache_checkout_key that redirects to the WooPay checkout if the user is logged in.
✅ Existing flows work as expected.
✅ Tested both in blocks and shortcode cart.
✅ Tested with a user logged in to the merchant store.
✅ Tested in sandbox with Jetpack blog auth.

bborman22 committed:
Add comment to explain conditional on minimum session

bborman22 committed:
Undo unrelated E2E fix

bborman22 committed:
Fix class docblock for WooPay session REST endpoint

bborman22 committed:
Remove outdated todo in WooPay session endpoint

bborman22 committed:
Update url parameter from direct checkout

ricardo reviewed (approved):
LGTM and tests well.

bborman22 committed:
Move encrypted session data to utility function and use it for direct checkout flow

bborman22 committed:
Add query parameters to redirect URL for direct checkout

bborman22 committed:
Remove references to cache session

bborman22 committed:
Fix merge conflicts

bborman22 commented:
@ricardo I have now updated the implementation to use an encrypted and signed session data passed through the redirect URL to WooPay instead of storing a temporary session on WooPay. This was based on my discussion with Systems and their recommendation that the storage could be unreliable in this use case. I have updated https://github.com/Automattic/woopay/pull/2529 as well to handle this implementation and so they should be tested again together. 

Sorry for the back and forth on this PR, but this should be the last reworking of the implementation now that we've had discussions more broadly.

ricardo commented on client/checkout/woopay/direct-checkout/woopay-direct-checkout.js:
```diff
@@ -291,6 +345,68 @@ class WooPayDirectCheckout {
+	 *
+	 * @return {boolean} True if all required fields are present, false otherwise.
+	 */
+	static validateEncryptedSessionData( encryptedSessionData ) {
+		return (
+			encryptedSessionData?.blog_id &&
+			encryptedSessionData?.data?.session &&
+			encryptedSessionData?.data?.iv &&
+			encryptedSessionData?.data?.hash
+		);
+	}
```
There's a very similar method in this file – `isValidEncryptedSessionData`.

Interesting how we sometimes end up writing almost the same code in separate PRs without knowing. 😅

bborman22 replied:
😆 Wow, can't believe I missed that! Now using just the existing function and removed the duplicate in https://github.com/Automattic/woocommerce-payments/pull/8268/commits/4b1b7d4c320e178232d3c1ebea1cfed829d7c355

ricardo commented on includes/woopay/class-woopay-session.php:
```diff
@@ -556,6 +532,54 @@ public static function ajax_get_woopay_session() {
+	 * 
+	 * @return array Array of minimum session data used by WooPay or false on failures.
+	 */
+	public static function get_woopay_minimum_session_data() {
+		$blog_id = Jetpack_Options::get_option('id');
+		if ( empty( $blog_id ) ) {
+			return [];
+		}
+
+		$data = [
+			'blog_id'           => $blog_id,
+			'blog_rest_url'     => get_rest_url(),
+			'blog_checkout_url' => wc_get_checkout_url(),
+			'session_nonce'     => self::create_woopay_nonce( get_current_user_id() ),
+			'store_api_token'   => self::init_store_api_token(),
+		];
+
+		return WooPay_Utilities::encrypt_and_sign_data( $data );
+	}
```
To prevent issues related to `openssl` not being enabled on the host, I guess we should add a `WC_Payments_Features::is_client_secret_encryption_eligible()` check before `encrypt_and_sign_data`, similar to how it's done in `get_frontend_init_session_request`.

bborman22 replied:
Great catch, I've copied the same check here now in https://github.com/Automattic/woocommerce-payments/pull/8268/commits/4b1b7d4c320e178232d3c1ebea1cfed829d7c355.

ricardo reviewed (changes requested):
Thank you for all the thought and effort on this new approach. It indeed looks better than the previous cache idea. 🚀 

I left 2 minor comments, but overall LGTM and tests well! 🎉

bborman22 committed:
Add encryption check and remove duplicate function in direct checkout

ricardo reviewed (approved):
LGTM!

bborman22 committed:
Merge branch 'develop' into add/woopay-merchant-request
